keps/sig-network/20200127-disable-service-type-lb-nodeports.md

```diff
@@ -0,0 +1,117 @@
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+As of today, a Kubernetes Service with Type=LoadBalancer will always allocate a node port. Though most implementations of Service Type=LoadBalancer do require node ports, there are several implementations that do not. This KEP proposes to add a new flag on the apiserver `--disable-loadbalancer-node-ports` to optionally disable node port allocation for Service Type=LoadBalancer.
+
+
+## Motivation
+
+There are several implementations of Service Type=LoadBalancer API today that do not consume the node ports automatically allocated by Kubernetes. Some examples include but are not limited to:
+* [MetalLB](https://github.com/danderson/metallb)
+* [kube-router](https://github.com/cloudnativelabs/kube-router)
+
+The drawbacks of allocating unused node ports are that:
+* the number of load balancers is now limited to the number of available node ports
+* node ports are exposed for an LB even though they are not used -- unecessarily exposed ports can fail regulatory/compliance requirements
+
+For clusters that have integrations for Service Type=LoadBalancer but don't require node ports should have the option to disable node port allocation.
+
+### Goals
+
+* disable node port for Service Type=LoadBalancer
+* don't change any existing default behavior for Service Type=LoadBalancer
+
+## Proposal
+
+A flag `--disable-loadbalancer-node-ports` will be added to the apiserver to disable node port allocations for Service Type=LoadBalancer. The flag's default value is `false` to ensure existing behavior is preserved.
+
+If `--disable-loadbalancer-node-ports` is `true`,  node port allocation/release for Service Type=LoadBalancer is skipped for all rest operations. If a Service Type=LoadBalancer already exists prior to the flag being set, the apiserver will still handle releasing its node port on delete.
+
+## Design Details
+
```

Comment:
What happens if I manually specify a NodePort (allowed today) - will it be ignored or respected?

Reply:
> Can you talk about why a global flag vs a per-service field or annotation + admission control vs ... other options? Global flags are scary.

Global flags made the most sense to me since implementations of service loadbalancer are also at a global level -- I'll add this.

 > What happens if I manually specify a NodePort (allowed today) - will it be ignored or respected?

It should be respected, we're only disabling automatic allocation. I'll update